### PR TITLE
fix: handle undefined typeParameters

### LINKED ISSUE #1427: TypeError: Cannot read properties of undefined (reading '0') - weird errors
### 🐛 Bug Report

-   TypeStat version: 0.7.2
-   TypeScript version: 4.9.5
-   Node version: 20.10.0

#### Actual Behavior

There are a lot errors like this when I try to run the tool

```sh
Error in <project_path>/<path>/<file>.ts at node 'export interface ApiResponse {
  deleted: number;
  errors: string[];
}' (position 1652):
	TypeError: Cannot read properties of undefined (reading '0')
	    at Object.expandReferencesForGenericTypes (<project_path>/node_modules/typestat/src/mutators/builtIn/fixIncompleteTypes/fixIncompleteInterfaceOrTypeLiteralGenerics/collectGenericNodeReferences.js:79:77)
	    at Object.collectGenericNodeReferences (<project_path>/node_modules/typestat/src/mutators/builtIn/fixIncompleteTypes/fixIncompleteInterfaceOrTypeLiteralGenerics/collectGenericNodeReferences.js:35:20)
	    at visitInterfaceOrTypeLiteral (<project_path>/node_modules/typestat/src/mutators/builtIn/fixIncompleteTypes/fixIncompleteInterfaceOrTypeLiteralGenerics/index.js:35:66)
	    at tryGetMutation (<project_path>/node_modules/typestat/src/mutators/collectMutationsFromNodes.js:46:16)
	    at visitNode (<project_path>/node_modules/typestat/src/mutators/collectMutationsFromNodes.js:33:30)
	    at visitNodes (<project_path>/node_modules/typescript/lib/typescript.js:31216:30)
	    at forEachChildInSourceFile (<project_path>/node_modules/typescript/lib/typescript.js:31575:20)
	    at Object.forEachChild (<project_path>/node_modules/typescript/lib/typescript.js:31990:47)
	    at Object.collectMutationsFromNodes (<project_path>/node_modules/typestat/src/mutators/collectMutationsFromNodes.js:40:8)
	    at fixIncompleteInterfaceOrTypeLiteralGenerics (<project_path>/node_modules/typestat/src/mutators/builtIn/fixIncompleteTypes/fixIncompleteInterfaceOrTypeLiteralGenerics/index.js:30:94)
```

Also, some places imports are overwritten like this. Not sure if these are connected. They do not seem to, but you never know.

```diff
-import {SearchApiService} from '../services/search-api.service';
+import {SearchApiService} from '../service | List<T>s/search-api.service';
```

#### Expected Behavior

Not throw weirds errors without explanation.

#### Reproduction

I have been trying with few different setups, but  here is latest

```json
{
  "include": [
    "src/**/*.ts"
  ],
  "filters": [
    "CallExpression[expression.text=suiteTeardown]",
    "CallExpression[expression.text=teardown]",
    "MethodDeclaration[name.text=dispose]"
  ],
  "fixes": {
    "incompleteTypes": true
  },
  "types": {
    "matching": [
      "^(null|undefined)$"
    ],
    "onlyPrimitives": true,
    "strictNullChecks": true
  }
}
```

Unfortunately, this is not a repo I can share straight away. It's Angular project, so there are legacy/experimental decorators for example. Strict options look like this

```json
    "experimentalDecorators": true,
    "alwaysStrict": true,
    "noImplicitAny": false,
    "noImplicitOverride": false,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "strictBindCallApply": true,
    "strictFunctionTypes": true,
    "strictNullChecks": false,
    "strictPropertyInitialization": false,
    "useUnknownInCatchVariables": true
```

I think this code that the error points to [this](https://github.com/JoshuaKGoldberg/TypeStat/blob/bdec7864dfd9563f1f5bfeeefcbdcb1990334879/src/mutators/builtIn/fixIncompleteTypes/fixIncompleteInterfaceOrTypeLiteralGenerics/collectGenericNodeReferences.ts#L78C1-L82C1).

```ts
        // Find the index of the corresponding type argument in the instantation of the node,
        // and finally use that to grab the node for the template parameter from the type argument's index
        const typeArgumentIndex = templatedParentInstantiation.typeArguments.indexOf(referencingNode);
        const templateDeclaredTypeNode = templatedDeclaration.typeParameters[typeArgumentIndex];
```

---

I have tried to restrict to run only on some files, and I suspect that Angular decorators are one culprit here.

### PR DESCRIPTION
**Pull request by rubiesonthesky**
## PR Checklist

- [x] Addresses an existing open issue: fixes #1427 
- [x] That issue was marked as [`status: accepting prs`](https://github.com/JoshuaKGoldberg/TypeStat/issues?q=is%3Aopen+is%3Aissue+label%3A%22status%3A+accepting+prs%22)
- [x] Steps in [CONTRIBUTING.md](https://github.com/JoshuaKGoldberg/TypeStat/blob/main/.github/CONTRIBUTING.md) were taken 🐙 

## Overview

`templatedDeclaration.typeParameters` was undefined in some situations, which threw an error.  Added new regression test category as this is not really test about how the fixes work otherwise. Probably the test case could be simplified, but I'm going with mindset that fixed bug is better than none.

### CODE REVIEW

File: test/cases/regresssionTests/1427/original.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+				results.push(...array);
+			}
+		}
+
+		return results;
+	};
+
+	const isNotUndefined = <T>(item: T | undefined): item is T =>
+		item !== undefined;
+})();
```
Comment (JoshuaKGoldberg) [inline test/cases/regresssionTests/1427/original.ts:80]: > not really test about how the fixes work otherwise

I think it is 😄. It's just testing an edge case or two that wasn't accounted for previously. A lot of the tests were added in as "regression" tests, but are now named generally.

I filed #1504 just now, too, for splitting up tests.

For this one, we can reproduce the same issue with a more targeted snippet:

```suggestion
(function () {
	interface Shape {
		value?: unknown;
	}

	[].reduce<Shape>(() => ({}), {});
})();
```

...which makes me think this could be named, maybe... `test/cases/fixes/incompleteTypes/nonGenericInterfaceAsTypeArgument` or something like that?

Was there anything else you'd wanted to test in here too?

File: src/shared/nodeTypes.ts [hunk 2]
```diff
@@ -65,7 +65,7 @@ export type NodeWithDefinedTypeArguments = ts.Node & {
 // TODO: make this a more specific type
 // Will have to deal with instantiations (new Container<T>() { ... }) and declarations (class Container<T>() { ... }))
 export type NodeWithDefinedTypeParameters = ts.Node & {
```
Comment (JoshuaKGoldberg) [inline src/shared/nodeTypes.ts:67]: [Naming] Nit: the original intent for this was that it's a "node" whose "type parameters" are _defined_. But now it's just a node with the property.

```suggestion
export type NodeWithTypeParameters = ts.Node & {
```

(here and the `isNodeWithDefinedTypeParameters` function too)

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
🔥 Nice find! The functionality makes sense to me. Fun little quirk of type predicates that you can be bit like this!

Requesting changes on a small nit for naming and a larger thing for shrinking down the test. Thanks!
